eessi/testsuite/tests/apps/osu.py

```diff
@@ -0,0 +1,296 @@
+"""
+This module tests the binary 'osu' in available modules containing substring 'OSU-Micro-Benchmarks'.
+The basic application class is taken from the hpctestlib to which extra features are added.
+"""
+
+import os
+import reframe as rfm
+from reframe.core.meta import parameters
+import reframe.utility.sanity as sn
+
+from hpctestlib.microbenchmarks.mpi.osu import osu_benchmark
+
+from eessi.testsuite import hooks, utils
+from eessi.testsuite.constants import SCALES, TAGS, DEVICE_TYPES
+
+def my_filtering_function():
+    """
+    Filtering function for filtering scales for the pt2pt OSU test
+    """
+    scale_filtered = SCALES.copy()
+    for key in list(SCALES):
+        if(key == '1_core' or key == '4_cores' or
+           SCALES.get(key).get('num_nodes') > 2):
+            scale_filtered.pop(key)
+        elif('node_part' in SCALES.get(key)):
+            if(SCALES.get(key).get('node_part') > 1):
+                scale_filtered.pop(key)
+    return scale_filtered
+
+
+def my_filtering_function_coll():
+    """
+    Filtering function for filtering scales for collective the OSU test
+    """
+    scale_filtered = SCALES.copy()
+    for key in list(SCALES):
+        if(key == '1_core'):
+            scale_filtered.pop(key)
+    return scale_filtered
+
+
+@rfm.simple_test
+class osu_pt_2_pt(osu_benchmark):
+    ''' Run-only OSU test '''
+    scale = parameter(my_filtering_function())
+    valid_prog_environs = ['default']
+    valid_systems = []
+    time_limit = '30m'
+    module_name = parameter(utils.find_modules('OSU-Micro-Benchmarks'))
+    # Device type for non-cuda OSU-Micro-Benchmarks should run on hosts of both
+    # node types. To do this the default device type is set to GPU.
+    device_type = DEVICE_TYPES['GPU']
+
+
+    @run_after('init')
+    def run_after_init(self):
+        """hooks to run after init phase"""
+        hooks.filter_valid_systems_by_device_type(
+               self,
+               required_device_type=self.device_type)
+        is_cuda_module = utils.is_cuda_required_module(self.module_name)
+        # This part of the hook is meant to be for the OSU cpu tests.
+        if not is_cuda_module and self.device_type == DEVICE_TYPES['GPU']:
+            self.valid_systems = ['*']
+            self.device_buffers = 'cpu'
```

Comment:
```suggestion
        if self.device_type == DEVICE_TYPES['CPU']:
            self.device_buffers = 'cpu'
```

Reply:
There is one condition missing in the `hooks` if it is not a `cuda` module and the device type is `GPU`. With OSU non cuda module one can run `cpu` tests in the `gpu` partition.  This is specific to this test and I have added a comment.
The if condition you have suggested has been incorporated later.

Reply:
@satishskamath running non-cuda modules on gpu partitions is decided by `filter_valid_systems_by_device_type` in combination with the config file:
- if a partition has feature FEATURES[CPU], then non-cuda modules can run on it
- if a partition has feature FEATURES[GPU], then cuda modules can run on it